### PR TITLE
[zh-cn]sync blog 2022-09-07-iptables-chains.md

### PR DESCRIPTION
**Pull request by my-git9**
[zh-cn]sync blog 2022-09-07-iptables-chains.md

### CODE REVIEW

File: content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md [hunk 1]
```diff
@@ -0,0 +1,343 @@
+---
+layout: blog
+title: "Kubernetes’s IPTables Chains Are Not API"
```
Comment (windsonsea) [inline content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md:3]:
```suggestion
title: "Kubernetes 的 iptables 链不是 API"
```

File: content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+或者这些链将包含任何特定的规则（即使它们确实存在）。
+
+<!--
+Then, in future releases, as part of [KEP-3178], we will begin phasing
+out certain chains that Kubernetes itself no longer needs. Components
+outside of Kubernetes itself that make use of `KUBE-MARK-MASQ`,
+`KUBE-MARK-DROP`, or other Kubernetes-generated iptables chains should
+start migrating away from them now.
+
+[KEP-3178]: https://github.com/kubernetes/enhancements/issues/3178
```
Comment (windsonsea) [inline content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md:54]:
```suggestion
[KEP-3178]: https://github.com/kubernetes/enhancements/issues/3178
```
Line 54 can be removed if not translated

File: content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+<!--
+## Background
+
+In addition to various service-specific iptables chains, kube-proxy
+creates certain general-purpose iptables chains that it uses as part
+of service proxying. In the past, kubelet also used iptables for a few
+features (such as setting up `hostPort` mapping for pods) and so it
+also redundantly created some of the same chains.
+-->
+## 背景
```
Comment (windsonsea) [inline content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md:71]:
```suggestion
## 背景   {#background}

```
Add a blank line after heading. If necessary, provide an anchor.

File: content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+creating these rules in the future.
+
+[the removal of dockershim]: https://kubernetes.io/blog/2022/02/17/dockershim-faq/
+[logs a warning]: https://access.redhat.com/solutions/6739041
+-->
+尽管在 Kubernetes 1.25，iptables kube-proxy 仍然很流行，
+并且 kubelet 继续创建它过去创建的 iptables 规则（尽管不再**使用**它们），
+第三方软件不能假设核心 Kubernetes 组件将来会继续创建这些规则。
+
+[移除dockershim]: https://kubernetes.io/blog/2022/02/17/dockershim-faq/
```
Comment (windsonsea) [inline content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md:113]:
```suggestion
[移除 dockershim]: https://kubernetes.io/zh-cn/blog/2022/02/17/dockershim-faq/
```

File: content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+  - `KUBE-MARK-MASQ`
+  - `KUBE-POSTROUTING`
+
+此外，`filter` 表中的 `KUBE-FIREWALL` 链将不再具有当前与
+`KUBE-MARK-DROP` 关联的功能（并且它最终可能会完全消失）。
+
+此更改将通过 `IPTablesOwnershipCleanup` 特性门控逐步实施。
+可以手动在 Kubernetes 1.25 开启此特性中进行测试。
+目前的计划是将其在 Kubernetes 1.27 中默认启用，
+尽管这可能会延迟到以后的版本。（不会在 Kubernetes 1.27 版本之前调整。）
```
Comment (windsonsea) [inline content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md:151]: Line 116 - 151 is too long. Please divide it into two or three segments, to provide convenience for reviewers.

File: content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+<!--
+This change will be phased in via the `IPTablesOwnershipCleanup`
+feature gate.  That feature gate is available and can be manually
+enabled for testing in Kubernetes 1.25. The current plan is that it
+will become enabled-by-default in Kubernetes 1.27, though this may be
+delayed to a later release. (It will not happen sooner than Kubernetes
+1.27.)
+-->
+此更改将通过 `IPTablesOwnershipCleanup` 特性门控逐步实施。
+可以手动在 Kubernetes 1.25 开启此特性中进行测试。
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md:148]:
```suggestion
你可以手动在 Kubernetes 1.25 中开启此特性进行测试。
```

File: content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+## What to do if you use Kubernetes’s iptables chains
+
+(Although the discussion below focuses on short-term fixes that are
+still based on iptables, you should probably also start thinking about
+eventually migrating to nftables or another API).
+-->
+## 如果你使用 Kubernetes 的 iptables 链怎么办
+
+（尽管下面的讨论侧重于仍然基于 iptables 的短期修复，
+但你可能也应该开始考虑最终迁移到 nftables 或其他 API）
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md:162]:
```suggestion
但你可能也应该开始考虑最终迁移到 nftables 或其他 API。）
```

File: content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+If you are making use of the `KUBE-MARK-MASQ` chain to cause packets
+to be masqueraded, you have two options: (1) rewrite your rules to use
+`-j MASQUERADE` directly, (2) create your own alternative “mark for
+masquerade” chain.
+-->
+### 如果你使用 `KUBE-MARK-MASQ` 链...  {#use-case-kube-mark-drop}
+
+如果你正在使用 `KUBE-MARK-MASQ` 链来伪装数据包，
+你有两个选择：（1）重写你的规则以直接使用 `-j MASQUERADE`，
+（2）创建你自己的替代方案 “mark for masquerade” 链。
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md:176]:
```suggestion
（2）创建你自己的替代链，完成“为伪装而设标记”的任务。
```

File: content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+then calls `-j KUBE-MARK-MASQ` to set a bit on the kernel packet mark
+as a reminder to itself. Then later, during `POSTROUTING`, it has a
+single rule that matches all previously-marked packets, and calls `-j
+MASQUERADE` on them.
+-->
+理论上，kube-proxy 可以有一组规则来匹配 `PREROUTING`/`OUTPUT`
+中的数据包并调用 `-j DNAT`，然后有第二组规则来匹配 `POSTROUTING`
+中的相同数据包并调用 `-j MASQUERADE`。
+但是，为了提高效率，kube-proxy 只匹配了一次，在 `PREROUTING`/`OUTPUT` 期间调用 `-j DNAT`，
+然后调用 `-j KUBE-MARK-MASQ` 在内核上设置一个比特的数据包标记该数据包作为对自身的提醒。
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md:208]:
```suggestion
然后调用 `-j KUBE-MARK-MASQ` 在内核数据包标记属性上设置一个比特，作为对自身的提醒。
```

File: content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+然后，在 `POSTROUTING` 期间，通过一条匹配所有先前标记的数据包的规则，对它们调用 `-j MASQUERADE`。
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md:209]:
```suggestion
然后，在 `POSTROUTING` 期间，通过一条规则来匹配所有先前标记的数据包，并对它们调用 `-j MASQUERADE`。
```

File: content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+want a similar arrangement. But in many cases, components that use
+`KUBE-MARK-MASQ` are only doing it because they copied kube-proxy’s
+behavior without understanding why kube-proxy was doing it that way.
+Many of these components could easily be rewritten to just use
+separate DNAT and masquerade rules. (In cases where no DNAT is
+occurring then there is even less point to using `KUBE-MARK-MASQ`;
+just move your rules from `PREROUTING` to `POSTROUTING` and call `-j
+MASQUERADE` directly.)
+-->
+如果你有**很多**规则需要像 kube-proxy 一样对同一个数据包应用 DNAT 和伪装，
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md:223]:
```suggestion
如果你有**很多**规则需要像 kube-proxy 一样对同一个数据包同时执行 DNAT 和伪装操作，
```

File: content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+`filter` table, because there are no complicated interactions between
+DNAT rules and drop rules, and so the drop rules can simply be moved
+from `nat` to `filter`.
+
+In more complicated cases, it might be necessary to “re-match” the
+same packets in both `nat` and `filter`.
+-->
+通常，删除对 `KUBE-MARK-DROP` 的依赖的方法与删除对 `KUBE-MARK-MASQ` 的依赖的方法相同。
+在 kube-proxy 的情况下，很容易将 `nat` 表中的 `KUBE-MARK-DROP` 
+的用法替换为直接调用 `filter` 表中的 `DROP`，因为它们之间没有复杂的交互 DNAT 规则和 DROP 规则，
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md:259]:
```suggestion
的用法替换为直接调用 `filter` 表中的 `DROP`，因为 DNAT 规则和 DROP 规则之间没有复杂的交互关系，
```

File: content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+`KUBE-MARK-DROP` in the `nat` table with direct calls to `DROP` in the
+`filter` table, because there are no complicated interactions between
+DNAT rules and drop rules, and so the drop rules can simply be moved
+from `nat` to `filter`.
+
+In more complicated cases, it might be necessary to “re-match” the
+same packets in both `nat` and `filter`.
+-->
+通常，删除对 `KUBE-MARK-DROP` 的依赖的方法与删除对 `KUBE-MARK-MASQ` 的依赖的方法相同。
+在 kube-proxy 的情况下，很容易将 `nat` 表中的 `KUBE-MARK-DROP`
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md:258]:
```suggestion
在 kube-proxy 的场景中，很容易将 `nat` 表中的 `KUBE-MARK-DROP` 
```

File: content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md [hunk 14]
```diff
@@ -0,0 +1,342 @@
...
+from `nat` to `filter`.
+
+In more complicated cases, it might be necessary to “re-match” the
+same packets in both `nat` and `filter`.
+-->
+通常，删除对 `KUBE-MARK-DROP` 的依赖的方法与删除对 `KUBE-MARK-MASQ` 的依赖的方法相同。
+在 kube-proxy 的情况下，很容易将 `nat` 表中的 `KUBE-MARK-DROP` 
+的用法替换为直接调用 `filter` 表中的 `DROP`，因为它们之间没有复杂的交互 DNAT 规则和 DROP 规则，
+因此 DROP 规则可以简单地从 `nat` 移动到 `filter`。
+更复杂的情况下，可能需要在 `nat` 和 `filter` 表中“重新匹配”相同的数据包。
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md:261]:
```suggestion
更复杂的场景中，可能需要在 `nat` 和 `filter` 表中“重新匹配”相同的数据包。
```

File: content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md [hunk 15]
```diff
@@ -0,0 +1,342 @@
...
+### If you use Kubelet’s iptables rules to figure out `iptables-legacy` vs `iptables-nft`... {#use-case-iptables-mode}
+
+Components that manipulate host-network-namespace iptables rules from
+inside a container need some way to figure out whether the host is
+using the old `iptables-legacy` binaries or the newer `iptables-nft`
+binaries (which talk to a different kernel API underneath).
+-->
+### 如果你使用 Kubelet 的 iptables 规则来确定 `iptables-legacy` 与 `iptables-nft`... {#use-case-iptables-mode}
+
+从容器内部操纵主机网络命名空间 iptables 规则的组件需要一些方法来确定主机是使用旧的
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md:273]:
```suggestion
对于从容器内部操纵主机网络命名空间 iptables 规则的组件而言，需要一些方法来确定主机是使用旧的
```

File: content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md [hunk 16]
```diff
@@ -0,0 +1,342 @@
...
+
+<!--
+The `iptables-wrappers` package has [already been updated] with this new
+heuristic, so if you were previously using that, you can rebuild your
+container images with an updated version of that.
+
+[`iptables-wrappers`]: https://github.com/kubernetes-sigs/iptables-wrappers/
+[already been updated]: https://github.com/kubernetes-sigs/iptables-wrappers/pull/3
+-->
+`iptables-wrappers` 包[已经更新]这个新的启发式，
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md:324]:
```suggestion
`iptables-wrappers` 包[已经被更新]，以提供这个新的启发式逻辑，
```

**Comment by mtardy** [comment]
Sorry for the failure it's because of my changes. If you want it to work:

1. [Synchronize your fork with upstream](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork)
1. Get the new changes on main: `git fetch && git checkout main && git pull origin main`
1. Rebase your branch on main `git checkout 2022-09-07-iptables-chains.md && git rebase -i main`

File: content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md [hunk 17]
```diff
@@ -0,0 +1,343 @@
+---
+layout: blog
+title: "Kubernetes 的 iptables 链不是 API"
+date: 2022-09-07
+slug: iptables-chains-not-api
+---
+
+<!--
+layout: blog
+title: "Kubernetes’s IPTables 链不是 API"
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md:10]: ？ 这里注释掉的英文原文为什么要翻译？

Comment (my-git9) [reply content/zh-cn/blog/_posts/2022-09-07-iptables-chains.md]: It was a mistake

**Review by ydFu — APPROVED** [review]
/lgtm

**Pull request merged**